service-worker/worker/src/plugins/dynamic/dynamic.ts

```diff
@@ -0,0 +1,127 @@
+import {FetchDelegate, FetchInstruction, Operation, Plugin, PluginFactory, UrlMatcher, NgSwCache, ScopedCache, VersionWorker, VersionWorkerImpl} from '@angular/service-worker/worker';
+import {DynamicGroup, ResponseWithSideEffect, DynamicStrategy, DynamicStrategyMap} from './group';
+import {DynamicManifest} from './manifest';
+
+export function Dynamic(strategies: DynamicStrategy[]): PluginFactory<DynamicImpl> {
+  return (worker: VersionWorker) => new DynamicImpl(worker as VersionWorkerImpl, strategies);
+}
+
+/**
+ * A plugin which implements dynamic content caching - the caching of requests to
+ * arbitrary URLs.
+ */
+export class DynamicImpl implements Plugin<DynamicImpl> {
+
+  /**
+   * The manifest configured by the user.
+   */
+  manifest: DynamicManifest;
+
+  /**
+   * All `DynamicGroup`s configured, each one representing a group defined in
+   * the manifest.
+   */
+  group: DynamicGroup[];
+
+  /**
+   * Map of `optimizeFor` strategies to their implementations.
+   */
+  strategies: DynamicStrategyMap = {};
+
+  /**
+   * `Promise` that tracks side effect application after requests have completed. This
+   * is used to serialize application of side effects, even if requests are executed
+   * in parallel.
+   */
+  sideEffectQueue: Promise<any>;
+
+  constructor(public worker: VersionWorkerImpl, strategies: DynamicStrategy[]) {
+    // Extract the dynamic section of the manifest.
+    this.manifest = worker.manifest['dynamic'];
+
+    // Initially there are no side effects.
+    this.sideEffectQueue = Promise.resolve();
+
+    // Build the `strategies` map from all configured strategies.
+    strategies.forEach(strategy => this.strategies[strategy.name] = strategy);
+  }
+
+  setup(ops: Operation[]): void {}
+
+  fetch(req: Request): FetchInstruction|null {
+    // If no dynamic caching configuration is provided, skip this plugin.
+    if (!this.manifest) {
+      return null;
+    }
+
+    // Return an instruction that applies dynamic content caching.
+    const instruction: FetchInstruction = (next: FetchDelegate): Promise<Response> => {
+      // There may be multiple groups configured. Check whether the request matches any
+      // of them.
+      const groups = this.group.filter(group => group.matches(req));
+      if (groups.length === 0) {
+        // It doesn't match any groups - continue down the chain.
+        return next();
+      }
+
+      // It has matched at least one group. Only the first group is considered.
+      return this
```

Comment:
No, it shouldn't cause any problems. Side effects are designed to run "down the line" instead of depending on happening immediately after the request which created them.